Always allow interim fields to be added to Analysis Services

ramonski opened the pull request:
## Description of the issue/feature this PR addresses

Interim fields in Analysis Services are the only way, how to capture multiple results for one Analysis, e.g. 5 different areas in the sample like middle , right , left etc.

Linked issue: https://github.com/senaite/senaite.core/issues/751

## Current behavior before PR

Interim fields are not visible when no calculation was selected

## Desired behavior after PR is merged

Interim fields are always visible

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

ramonski commented:
## Welcome to the Jungle 🌴 🙈 🌴 

This PR refactors the whole JS logic in Analysis Services.

The following functionality is given:

- [ ] 
- [ ] 
- [ ] 
- [ ] 
- [ ]

xispa commented on bika/lims/browser/js/coffee/bika.lims.analysisservice.coffee:
```diff
@@ -1,726 +1,1326 @@
+      # {uid: "488400e9f5e24a4cbd214056e6b5e2aa", title: "My Calculation"}
+      if not @is_uid data.uid
+        return deferred.resolveWith this, [{}]
+
```
This "Instrument assignment is not required"/"Manual entry of results" thing is totally weird. It shouldn't rely on the assignment/unassignment of methods, rather with instruments (if no instrument assigned, then there is no choice, the user *must* set the result manually)... but in turn, Method content type has the option "Manual entry" True/False, at the same time that can have none or multiple instruments associated.... so well, kind of an hydra we'll have to deal with in future.

ramonski commented on bika/lims/browser/js/coffee/bika.lims.analysisservice.coffee:
```diff
@@ -1,726 +1,1326 @@
+    # load the calculation now, to set the interims
+    @load_calculation uid
+    .done (calculation) ->
+      @set_interims calculation.InterimFields
```
Manual interims are always preserved, see `set_interims`.

However, I fixed a small issue, where "overwritten" interims, e.g. manual interims with the same keyword as the calculation interim, had different values, e.g. `value` or `unit` etc.
These interims are now merged on calculation change and preserved as "manual" interim

ramonski commented on bika/lims/browser/js/coffee/bika.lims.analysisservice.coffee:
```diff
@@ -1,726 +1,1326 @@
+    deferred = $.Deferred()
+
+    options =
+      url: @get_portal_url() + "/get_available_calculations"
```
You are right. This method is a relict that remained from refactoring and nowhere used. I'm removing it...

ramonski commented on bika/lims/browser/js/coffee/bika.lims.analysisservice.coffee:
```diff
@@ -1,726 +1,1326 @@
+      data:
+        catalog_name: "bika_setup_catalog"
+        page_size: 0
+        portal_type: "Instrument"
```
Indeed, I added issue https://github.com/senaite/senaite.core/issues/829 to not forget about this

ramonski commented on bika/lims/browser/js/coffee/bika.lims.analysisservice.coffee:
```diff
@@ -1,726 +1,1326 @@
+        page_size: 0
+        portal_type: "Method"
+        inactive_state: "active"
+        sort_on: "sortable_title"
```
This one was not used, so I removed it from the JS

ramonski commented:
Hi Jordi, thanks for your review and feedback.
I implemented the changes that you suggested and fixed a bug as well.

Could you please functional check the following:

- Create 2 Calculations with the interim field `A` and `B` respectively
- Assign calculation `A` to AS (Interim `A` should be added automatically)
- Change the `value` or `unit` of Interim `A` in the AS and save
- Switch in the AS  tab between the calculations `A` and `B`

Expected:

Interim `A` is always preserved on calculation change, but merged with the Calculation `A`'s interim

ramonski commented:
Hi Jordi, good catch. I added this functionality.
